File: qsort_ffi.lua
#!/usr/bin/env tarantool

local clock = require("clock")
local ffi = require("ffi")
local log = require("log")

ffi.cdef([[
    void quick_sort(int* arr, size_t left, size_t right);
]])

local mylib = ffi.load("./qsortf.so")

math.randomseed(os.time())

local array = { }
local n = 10 * 1000 * 1000
for i = 1, n do
    array[i] = math.random(1000000)
end

local start = tonumber(clock.proc64())
local arg = ffi.new(("int[%d]"):format(n), array)
mylib.quick_sort(arg, 0, n - 1)
local stop = tonumber(clock.proc64())

log.info("Time: %.2f sec", (stop - start) / 1000 / 1000 / 1000)

File: qsort_ffi.c
#define SWAP_INT(a, b) \
    { \
        int tmp = a; \
        a = b; \
        b = tmp; \
    }

static int partition(int* arr, int left, int right)
{
    int pivot_value = arr[right];
    int pivot_pos = left - 1;

    for (int i = left; i < right; i++)
        if (arr[i] < pivot_value) {
            pivot_pos++;
            SWAP_INT(arr[i], arr[pivot_pos]);
        }

    pivot_pos++;
    SWAP_INT(arr[right], arr[pivot_pos]);

    return pivot_pos;
}

void quick_sort(int* arr, int left, int right)
{
    if (left < right) {
        int pivot_pos = partition(arr, left, right);
        quick_sort(arr, left, pivot_pos - 1);
        quick_sort(arr, pivot_pos + 1, right);
    }
}
